packages/medusa/src/services/line-item.ts

```diff
@@ -178,115 +185,205 @@ class LineItemService extends TransactionBaseService {
+            lineItem.adjustments =
+              adjustments as unknown as LineItemAdjustment[]
+          }
+
+          generatedItems.push(lineItem)
         }
 
-        return lineItem
+        return (isDataAnArray
+          ? generatedItems
+          : generatedItems[0]) as unknown as TResult
       }
     )
   }
 
+  protected async generateLineItem(
+    variant: {
+      id: string
+      title: string
+      product_id: string
+      product: {
+        title: string
+        thumbnail: string | null
+        discountable: boolean
+        is_giftcard: boolean
+      }
+    },
+    quantity: number,
+    context: GenerateLineItemContext & {
+      variantPricing: ProductVariantPricing
+    }
+  ): Promise<LineItem> {
+    const transactionManager = this.transactionManager_ ?? this.manager_
+
+    let unit_price = Number(context.unit_price) < 0 ? 0 : context.unit_price
+    let unitPriceIncludesTax = false
+    let shouldMerge = false
+
+    if (context.unit_price == null) {
+      shouldMerge = true
+
+      unitPriceIncludesTax =
+        !!context.variantPricing?.calculated_price_includes_tax
+      unit_price = context.variantPricing?.calculated_price ?? undefined
+    }
+
+    const rawLineItem: Partial<LineItem> = {
+      unit_price: unit_price,
+      title: variant.product.title,
+      description: variant.title,
+      thumbnail: variant.product.thumbnail,
+      variant_id: variant.id,
+      quantity: quantity || 1,
+      allow_discounts: variant.product.discountable,
+      is_giftcard: variant.product.is_giftcard,
+      metadata: context?.metadata || {},
+      should_merge: shouldMerge,
+    }
+
+    if (
+      this.featureFlagRouter_.isFeatureEnabled(
+        TaxInclusivePricingFeatureFlag.key
+      )
+    ) {
+      rawLineItem.includes_tax = unitPriceIncludesTax
+    }
+
+    if (this.featureFlagRouter_.isFeatureEnabled(OrderEditingFeatureFlag.key)) {
+      rawLineItem.order_edit_id = context.order_edit_id || null
+    }
+
+    const lineItemRepo = transactionManager.getCustomRepository(
+      this.lineItemRepository_
+    )
+
+    const lineItem = lineItemRepo.create(rawLineItem)
+    lineItem.variant = variant as ProductVariant
+
+    return lineItem
+  }
+
   /**
    * Create a line item
    * @param data - the line item object to create
    * @return the created line item
    */
-  async create(data: Partial<LineItem>): Promise<LineItem> {
+  async create<
+    T = LineItem | LineItem[],
+    TResult = T extends LineItem ? LineItem : LineItem[]
+  >(data: T): Promise<TResult> {
     return await this.atomicPhase_(
       async (transactionManager: EntityManager) => {
         const lineItemRepository = transactionManager.getCustomRepository(
           this.lineItemRepository_
         )
 
-        const item = lineItemRepository.create(data)
-        return await lineItemRepository.save(item)
+        const isDataAnArray = Array.isArray(data)
+        const data_ = !isDataAnArray ? [data] : data
```

Comment:
suggestion(non-blocking):
```suggestion
        const data_ = Array.isArray(data) ? data : [data]
```